community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java

```diff
@@ -163,37 +167,70 @@ public String getTypeDescriptor()
     }
 
     /**
-     * This method is called by constructors.
+     * This method is called by constructors. Checks the header record and loads the store.
+     * <p>
+     * Note: This method will map the file with the page cache. The store file must not
+     * be accessed directly until it has been unmapped - the store file must only be
+     * accessed through the page cache.
      * @param createIfNotExists If true, creates and initialises the store file if it does not exist already. If false,
      * this method will instead throw an exception in that situation.
      */
-    protected void checkStorage( boolean createIfNotExists )
+    protected void checkAndLoadStorage( boolean createIfNotExists )
     {
         int pageSize = pageCache.pageSize();
-        //noinspection EmptyTryBlock
-        try ( PagedFile ignore = pageCache.map( storageFileName, pageSize, ANY_PAGE_SIZE ) )
+        int filePageSize;
+        try ( PagedFile pagedFile = pageCache.map( storageFileName, pageSize, ANY_PAGE_SIZE ) )
         {
+            extractHeaderRecord( pagedFile );
+            filePageSize = pageCache.pageSize() - pageCache.pageSize() % getRecordSize();
         }
-        catch ( NoSuchFileException e )
+        catch ( NoSuchFileException | StoreNotFoundException e )
         {
             if ( createIfNotExists )
             {
-                try ( PagedFile file = pageCache.map( storageFileName, pageSize, StandardOpenOption.CREATE ) )
+                try
                 {
-                    initialiseNewStoreFile( file );
+                    createStore( pageSize );
                     return;
                 }
                 catch ( IOException e1 )
                 {
                     e.addSuppressed( e1 );
                 }
             }
+            if ( e instanceof StoreNotFoundException )
+            {
+                throw (StoreNotFoundException) e;
+            }
             throw new StoreNotFoundException( "Store file not found: " + storageFileName, e );
         }
         catch ( IOException e )
         {
             throw new UnderlyingStorageException( "Unable to open store file: " + storageFileName, e );
         }
+        loadStorage( filePageSize );
+    }
+
+    private void createStore( int pageSize ) throws IOException
+    {
+        try ( PagedFile file = pageCache.map( storageFileName, pageSize, StandardOpenOption.CREATE ) )
+        {
+            initialiseNewStoreFile( file );
+        }
+        checkAndLoadStorage( false );
```

Comment:
In order to read the header record, calculate the `filePageSize` and properly handle io errors.